# SCIKIT LEARN WORKFLOW
## END TO END SCIKIT LEARN WORKFLOW

What is end to end means? 

One thing to note though Scikit Learn is such big library thus only call what you need will be a better approach to practical use.

### Random Forest Classifier Workflow for Classifying Heart Disease.
#### 1. Get the data ready
I will use the heart disease data. Remember that it was stored in csv format. First I need to import it first. Pandas will be the appropriate library to get and process the data to form Data Frame.

In [29]:
# get the data heart-disease.csv
import pandas as pd
heart_disease_data = pd.read_csv('../data/heart-disease.csv')
heart_disease_data

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
0     63    1   3       145   233    1        0      150      0      2.3   
1     37    1   2       130   250    0        1      187      0      3.5   
2     41    0   1       130   204    0        0      172      0      1.4   
3     56    1   1       120   236    0        1      178      0      0.8   
4     57    0   0       120   354    0        1      163      1      0.6   
..   ...  ...  ..       ...   ...  ...      ...      ...    ...      ...   
298   57    0   0       140   241    0        1      123      1      0.2   
299   45    1   3       110   264    0        1      132      0      1.2   
300   68    1   0       144   193    1        1      141      0      3.4   
301   57    1   0       130   131    0        1      115      1      1.2   
302   57    0   1       130   236    0        0      174      0      0.0   

     slope  ca  thal  target  
0        0   0     1       1  
1        0   0     2       1  
2        2   0     2       1  
3        2   0     2       1  
4        2   0     2       1  
..     ...  ..   ...     ...  
298      1   0     3       0  
299      1   0     3       0  
300      1   2     3       0  
301      1   1     3       0  
302      1   1     2       0  

[303 rows x 14 columns]

***NOTE***
The heart_disease_data consist of all patient characteristics except for the target column. Target column is the end result whether the patient has heart_disease or not and we want to seek the connection with each character.

In [30]:
# first we need to separate the target column from other data
x_data = heart_disease_data.drop('target', axis=1)
x_data.head()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
2   41    0   1       130   204    0        0      172      0      1.4      2   
3   56    1   1       120   236    0        1      178      0      0.8      2   
4   57    0   0       120   354    0        1      163      1      0.6      2   

   ca  thal  
0   0     1  
1   0     2  
2   0     2  
3   0     2  
4   0     2

In [31]:
# now the target label will be in the y_data
y_data = heart_disease_data['target']
y_data.head()

0    1
1    1
2    1
3    1
4    1
Name: target, dtype: int64

***NOTE:***
now I need to split the data frame in the x_data and y_data to do this I will use the scikit learn libs

The way to import is a bit different since it is named sklearn thus to import I need import sklearn.

Now as this library is quite bid I need to be specific. This time it is model_selection to be exact I want to randomly split the data frame heart_disease_data y_data and x_data. I will use the train_test_split library

In [32]:
# import the library
# https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html
from sklearn.model_selection import train_test_split

# just to be safe if we need to handle some matrices I will import numpy as well
import numpy as np

In [33]:
x_train, x_test, y_train, y_test = train_test_split(x_data, y_data)
# let's test the result but since too big I will just validate each size
x_train.shape, y_train.shape, x_test.shape, y_test.shape

((227, 13), (227,), (76, 13), (76,))

***NOTE:***
The data (shape) above is important. 

I need to validatae that the x_train and y_train first size is the same (in this case 227).

The same case is true for the x_test and y_test (in this case 76)

It means will be 227 data for train and 76 for testing that chosen randomly.

#### 2. Choose the model and hyperparameters
Now this is the very heart of the Scikit Learn. A variety of models or in machine learning is known as classifiers. However, in Scikit documentation it is more commonly known as estimators. 

TODO So what is hyperparameters??

Next we will use Random Forest Classifier: https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html?highlight=forest#sklearn.ensemble.RandomForestClassifier

 learn about Random Forest Classifier model: https://en.wikipedia.org/wiki/Random_forest
 It is said that this is part of the ensemble learning method thus it is put in the Sklearn ensemble libs

In [34]:
# import the Random Forest 
from sklearn.ensemble import RandomForestClassifier
# too long the name thus better put it into a variable
clf = RandomForestClassifier()

In [35]:
clf.get_params()

{'bootstrap': True,
 'ccp_alpha': 0.0,
 'class_weight': None,
 'criterion': 'gini',
 'max_depth': None,
 'max_features': 'auto',
 'max_leaf_nodes': None,
 'max_samples': None,
 'min_impurity_decrease': 0.0,
 'min_impurity_split': None,
 'min_samples_leaf': 1,
 'min_samples_split': 2,
 'min_weight_fraction_leaf': 0.0,
 'n_estimators': 100,
 'n_jobs': None,
 'oob_score': False,
 'random_state': None,
 'verbose': 0,
 'warm_start': False}

#### 3. Fit the model to the data and use it to make predictions
Fitting the model on the data involves passing it the data and asking it to figure out the patterns.

If there are labels (supervised learning), the model tries to work out the relationship between the data and the labels.

If there are no labels (unsupervised learning), the model tries to find patterns and group similar samples together.

First I need to build the forest of trees model from the training set, in this case is **x_train and y_train.**

Read more about the forest of the random forest classifier's fit method: https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html#sklearn.ensemble.RandomForestClassifier.fit



In [36]:
# making random forest classifier 
clf.fit(x_train, y_train)

RandomForestClassifier()

***NOTE:***
This clf.fit does not need to be stored in variable. I think it will just use the clf object to store the data from both side of randomly selected train data. 


Here clf = RandomForestClassifier is an **object**. It is just like the **DataFrame** in Pandas. You can modify directly the instatiation of this DataFrame (in this case heart_disease_data) which you can cut, process, etc. 

#### Use the model to make a prediction
The whole point of training a machine learning model is to use it to make some kine of prediction in the future. 

Once our model instance is trained, you can use the predict() method to predict a target value given a set of features. In other words, use the model, along with some unlabelled data to predict the label.

***CAUTION***: data you predict on has to be in the same shape as data you trained on.

Fortunately we already have the correct shape since we built our train and test data randomly using the Scikit Learn **train_test_split()** function. This will ensure the x_train and x_test will be the same shape. (We already verified this above), emphasis in the number of column (in this case the second in the shape function output).

In [37]:
# just to verify
x_test

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
234   70    1   0       130   322    0        0      109      0      2.4   
33    54    1   2       125   273    0        0      152      0      0.5   
214   56    1   0       125   249    1        0      144      1      1.2   
138   57    1   0       110   201    0        1      126      1      1.5   
61    54    1   1       108   309    0        1      156      0      0.0   
..   ...  ...  ..       ...   ...  ...      ...      ...    ...      ...   
24    40    1   3       140   199    0        1      178      1      1.4   
166   67    1   0       120   229    0        0      129      1      2.6   
90    48    1   2       124   255    1        1      175      0      0.0   
247   66    1   1       160   246    0        1      120      1      0.0   
254   59    1   3       160   273    0        0      125      0      0.0   

     slope  ca  thal  
234      1   3     2  
33       0   1     2  
214      1   1     2  
138      1   0     1  
61       2   0     3  
..     ...  ..   ...  
24       2   0     3  
166      1   2     3  
90       2   2     2  
247      1   3     1  
254      2   0     2  

[76 rows x 13 columns]

In [38]:
# now using the predict method to make prediction onf the test data 
y_preds = clf.predict(x_test)

#### 4. Evaluate the model
Now we've made some predictions (y_preds). we can start to use some more Scikit-Learn methods to figure out how good our models is. 

Each model or estimator has built-in score method. This method compares how well the model was able to learn the patterns between the features and labels. In other words, it returns how accurate your model is.

Using Random Forest Classifier score method. 

**score(X, y[, sample_weight])** https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html#sklearn.ensemble.RandomForestClassifier.score

Return the mean accuracy on the given test data and labels.

In [39]:
clf.score(x_train, y_train)

1.0

In [40]:
clf.score(x_test, y_test)

0.8552631578947368

***NOTE:*** this numbers are the mean accuracy between two data set. After using the predictin model we can use the model on the test data sets. 

#### use the additional evaluation methods and also libs for reporting
This is part of the metrics libs of the sklearn. 

See the [Metrics and scoring: quantifying the quality of predictions](https://scikit-learn.org/stable/modules/model_evaluation.html#model-evaluation) section and the Pairwise metrics, Affinities and Kernels section of the user guide for further details.

The [sklearn.metrics module](https://scikit-learn.org/stable/modules/classes.html#module-sklearn.metrics) includes score functions, performance metrics and pairwise metrics and distance computations.

Here is the list I need to use from this lib:
1. [classification_report](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.classification_report.html#sklearn.metrics.classification_report)
1. [confusion_matrix](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.confusion_matrix.html#sklearn.metrics.confusion_matrix)
1. [accuracy_score](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.accuracy_score.html#sklearn.metrics.accuracy_score)


[3.3.2.7. Classification report](https://scikit-learn.org/stable/modules/model_evaluation.html#classification-report)

The classification_report function builds a text report showing the main classification metrics. Here is a small example with custom target_names and inferred labels:

**sklearn.metrics.classification_report(y_true, y_pred, *, labels=None, target_names=None, sample_weight=None, digits=2, output_dict=False, zero_division='warn')**[source](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.classification_report.html#sklearn.metrics.classification_report)

In [41]:
# import the sklearn.metric additiona evaluation and reporting methods
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score

# now print the classification report y_true is y_test, y_pred is well y_preds
print(classification_report(y_test, y_preds))

              precision    recall  f1-score   support

           0       0.82      0.89      0.85        35
           1       0.89      0.83      0.86        41

    accuracy                           0.86        76
   macro avg       0.86      0.86      0.86        76
weighted avg       0.86      0.86      0.86        76



From this lib I need [confusion matrix (here is the user guide on it)](https://scikit-learn.org/stable/modules/model_evaluation.html#confusion-matrix). Also here is [wikipedia page on cofusion matrix](https://en.wikipedia.org/wiki/Confusion_matrix).

**sklearn.metrics.confusion_matrix(y_true, y_pred, *, labels=None, sample_weight=None, normalize=None)**

returns: 
    C: ndarray of shape (n_classes, n_classes)

Confusion matrix whose i-th row and j-th column entry indicates the number of samples with true label being i-th class and predicted label being j-th class.

In [42]:
# use the confusion matrix with y_true = y_test and y_pred = y_preds
confusion_matrix(y_test, y_preds)

array([[31,  4],
       [ 7, 34]], dtype=int64)

Next is the [accuracy_score](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.accuracy_score.html#sklearn.metrics.accuracy_score)

**sklearn.metrics.accuracy_score(y_true, y_pred, *, normalize=True, sample_weight=None)**

Returns: score: float

If normalize == True, return the fraction of correctly classified samples (float), else returns the number of correctly classified samples (int).

The best performance is 1 with normalize == True and the number of samples with normalize == False.

*see the example in the source guide*

In [43]:
# use the accuracy_score with y_true = y_test and y_pred = y_preds
accuracy_score(y_test, y_preds)

0.8552631578947368

#### 5. Experiment to improve
The first model we build is often referred  to as baseline.

Once we have got a baseline model, like we have here, it is important to remember, this is often not the final model we will use. 

The next step in the workflow is to try to improve upon your baseline model.

To do this there are two ways to look at it: 
1. Form a model perspective 
1. From the data perspective

From model perspective maybe would involves things such as using a more complex model or tuning current model's hyper-parameters.

From data perspective, this may involves collecting more data or improving the data quality. This will give better chance of the model to learn the pattern within.

If you are already working on an existing dataset, it is often easier to try a series of model perspective experiments first then turn to data perspective if you are not getting the result you are looking for. This is sounds like our current case where we alreasy have heart disease data frame.

One thing you should be aware of is if you are tuning a model's hyper-parameters in a series of experiments, your results should always be **cross-validated**. 

Cross validation is a way of making sure that results you are getting are consistent across your training and test dataset. This must validate the use of multiple versions of training and test sets rather than just pure luck because of the order the training and test sets were created.
- try different hyper-parameters
- All different parameters should be cross validated
    - **CAUTION:** beware of cross validation for time series problems

Different models you use will have different hyper-parameters you can tune. For the case of our model the Random Forrest Classifier, we will start trying different values for n_estimators.

In [44]:
# Try different numbers of estimators (trees) ... (no cross-validation)
# why does it need the random seed?
# this is important since we will compare this results with the one with cross validation
np.random.seed(42)

# we will make estimators of (10, 20, 30, 40, 50, 60, 70, 80, 90)
for i in range(10,100,10):
    print(f"trying model with {i} estimators...")
    model = RandomForestClassifier(n_estimators=i).fit(x_train, y_train)
    print(f"Model accuracy on test: {model.score(x_test, y_test) * 100}%")
    print("")

trying model with 10 estimators...
Model accuracy on test: 76.31578947368422%

trying model with 20 estimators...
Model accuracy on test: 81.57894736842105%

trying model with 30 estimators...
Model accuracy on test: 80.26315789473685%

trying model with 40 estimators...
Model accuracy on test: 82.89473684210526%

trying model with 50 estimators...
Model accuracy on test: 82.89473684210526%

trying model with 60 estimators...
Model accuracy on test: 85.52631578947368%

trying model with 70 estimators...
Model accuracy on test: 84.21052631578947%

trying model with 80 estimators...
Model accuracy on test: 82.89473684210526%

trying model with 90 estimators...
Model accuracy on test: 88.1578947368421%



In [45]:
# now we will compare the different numbers of estimators but with cross validation
# import the coross validation score lib
from sklearn.model_selection import cross_val_score

# re-seeding the random generator and set the same number series
np.random.seed(42)

# rerun similar test but adding cross validation
for i in range(10,100,10):
    print(f"trying model with {i} estimators...")
    model = RandomForestClassifier(n_estimators=i).fit(x_train, y_train)
    print(f"Model accuracy on test: {model.score(x_test, y_test) * 100}%")
    print(f"Cross-validation score: {np.mean(cross_val_score(model, x_data, y_data, cv=5)) * 100}%")
    print("")

trying model with 10 estimators...
Model accuracy on test: 76.31578947368422%
Cross-validation score: 78.53551912568305%

trying model with 20 estimators...
Model accuracy on test: 81.57894736842105%
Cross-validation score: 79.84699453551912%

trying model with 30 estimators...
Model accuracy on test: 82.89473684210526%
Cross-validation score: 80.50819672131148%

trying model with 40 estimators...
Model accuracy on test: 84.21052631578947%
Cross-validation score: 82.15300546448088%

trying model with 50 estimators...
Model accuracy on test: 82.89473684210526%
Cross-validation score: 81.1639344262295%

trying model with 60 estimators...
Model accuracy on test: 82.89473684210526%
Cross-validation score: 83.47540983606557%

trying model with 70 estimators...
Model accuracy on test: 86.8421052631579%
Cross-validation score: 81.83060109289617%

trying model with 80 estimators...
Model accuracy on test: 85.52631578947368%
Cross-validation score: 82.81420765027322%

trying model with 90 estim

***NOTE***: the cross_val_score function return an array. Thus the score must be averaged in order to make some practical scale. Therefore, it is needed numpy means to average the cross_val_score return.

[learn more on cross_val_score](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.cross_val_score.html#sklearn.model_selection.cross_val_score)

question: why does it build their own model each time trying new amount of estimator? Can it just use the clf object made earlier and see the changes from there?

Another way to find the best estimator is using [GridSearchCV sklearn model](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.GridSearchCV.html#sklearn.model_selection.GridSearchCV).

In [18]:
# now we use GridSearchCV
from sklearn.model_selection import GridSearchCV

# re-seeding the random number generator
np.random.seed(42)

# the main part for the grid search is defining the param_grid
# param_grid is a dict type with str as key and list as value
# the key (str) is the parameter of the estimator to be iterate and tested to find the most optimal parameter
param_grid = {'n_estimators': [i for i in range(10,100,10)]}

# preparing the grid
grid = GridSearchCV(RandomForestClassifier(),param_grid, cv=5)

# use the grid to fit the whole data
grid.fit(x_data, y_data)

# find the best grid
grid.best_params_

{'n_estimators': 80}

In [19]:
# now we use the parameter in the grid.best_params_ to be used to the selected estimator (in this case Random Forest Classifier)
# then set the parameter on n_estimator in that estimator using the grid.best_params_
# then we set our clf (we set earlier in our first attempt to model the whole data) according to the best_estimator_ 
clf = grid.best_estimator_
clf

RandomForestClassifier(n_estimators=80)

In [20]:
# use the best model to fit
clf = clf.fit(x_train, y_train)

In [21]:
# rate the score for the new fit model
clf.score(x_test,y_test)

0.7631578947368421

#### 6. Save a model for someone else to use
When you've done a few experiments and you're happy with how your model is doing, you'll likely want someone else to be able to use it.

This may come in the form of a teammate or colleague trying to replicate and validate your results or through a customer using your model as part of a service or apllication you offer.

Saving a model also allows you to reuse it later without having to go through retrianing it. Which is helpful, especially when your training times start to increase. 

You can save a scikit-learn model using [Python's built-in pickle module](https://docs.python.org/3/library/pickle.html) which serialize objects.

In this session I will try to serialize model which tried in number of estimator. Note that the last model is made the number of estimator of 90.

The score for the last model with number of estimator of 90 is 86.8421052631579% which we will use to validate the model once it was re-loaded

In [22]:
# we use pickle to serialize the model (only the model not the whole notebook)
import pickle

# save the model called model (the one iterated for number of estimator 10 to 100)
pickle.dump(model, open('random_forest_model_1.pkl', 'wb'))

***NOTE***: After this command is run you can confirm that random_forest_model_1.pkl file is created and available in the same folder as this ipynb file location.

Let's validate by loading the model form the random_forest_model_1.pkl file

In [23]:
loaded_model = pickle.load(open('random_forest_model_1.pkl', 'rb'))
loaded_model

RandomForestClassifier(n_estimators=90)

***NOTE***: The loaded model is the same type and having the same n_estimators parameter as the latest model (which is 90). But we still need to verify the model to be consistent which means it must have the same performace score as the orignal model.

In [24]:
# testing the loaded_model consistency
# the score should match the model(n_estimator=90) score which is 86.8421052631579%
loaded_model.score(x_test, y_test)

0.7763157894736842

***NOTE***: the score when testing the x_test and y_test is 0.868421052631579 which is the same as 86.8421052631579%.

Meaning the loaded_model is consistent as the serialized original model(n_estimators=90)

### 1. Getting the data ready
Data doesn't always come ready to use with Scikit-Learn machine learning model.

Three of the main steps you'll often have to take are:
- [x] Splitting the data into features (usually X) and labels (usually y)
- [] Filling (also called **imputing**) or disregarding missing values
- [] Converting non-numerical values to numerical values (also called **encoding**)

Let's see an example. We going back to the heart disease data above. 

In [25]:
heart_disease_data.head()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
2   41    0   1       130   204    0        0      172      0      1.4      2   
3   56    1   1       120   236    0        1      178      0      0.8      2   
4   57    0   0       120   354    0        1      163      1      0.6      2   

   ca  thal  target  
0   0     1       1  
1   0     2       1  
2   0     2       1  
3   0     2       1  
4   0     2       1

#### 1.1. Make sure the data are all numerical
To see this we need to shift to using the Car Sales DataFrame.

Many of the data in Car Sales DataFrame are non numerical data. We need to make it numeric by changing its type or giving a numerical representation if the data is ordinal.

In [26]:
# import the car-sales-extended.csv
car_sales = pd.read_csv("../data/car-sales-extended.csv")
car_sales

Make Colour  Odometer (KM)  Doors  Price
0     Honda  White          35431      4  15323
1       BMW   Blue         192714      5  19943
2     Honda  White          84714      4  28343
3    Toyota  White         154365      4  13434
4    Nissan   Blue         181577      3  14043
..      ...    ...            ...    ...    ...
995  Toyota  Black          35820      4  32042
996  Nissan  White         155144      3   5716
997  Nissan   Blue          66604      4  31570
998   Honda  White         215883      4   4001
999  Toyota   Blue         248360      4  12732

[1000 rows x 5 columns]

In [27]:
# check the data types in the car_sales dataFrame
car_sales.dtypes

Make             object
Colour           object
Odometer (KM)     int64
Doors             int64
Price             int64
dtype: object

***NOTE***:
The data types are consist of int64 which is a number and object which most likely is not a number. This will be used also as measurement for model. 

Now as in the previous model in heart disease I need to separate the data into X which is the data and y the target data in the model. In this case the target (y) is the Price data and X are the rest.

In [50]:
# the data target is y price
car_sales_X = car_sales.drop("Price", axis=1)
# axis = 0 means row, axis=1 means column
car_sales_y = car_sales["Price"]

# now we split the data into train and test
car_sales_X_train, car_sales_X_test, car_sales_y_train, car_sales_y_test = train_test_split(car_sales_X, car_sales_y, test_size=0.2)

***NOTE***: The data Colour and Make are non numeric types, this cannot be used directly for building models to predict the other data.

I want to find model to predict the price as function of other data in the data frame. Thus in the code above car_sales_y is set on price. Then the car_sales_X set to drop Price column.

I need to use another [Random Forest Regressor](https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestRegressor.html#)

We will use sklearn.preprocessing OneHotEncoder ; [read documentation](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.OneHotEncoder.html)

This package used to encode categorical features as a one-hot numeric array. This is used to process discrete or categorical feature. NOTE FEATURE NOT DATA! After we process the feature then we grouped the data as specified.

The next package I need is the column transformer. Here is the [link to the documentation](https://scikit-learn.org/stable/modules/generated/sklearn.compose.ColumnTransformer.html). This closely related to the pandas DataFrame. Basically this package will change certain column(s) into a specified features.

In [51]:
car_sales_X_train.shape, car_sales_y_train.shape, car_sales_X_test.shape, car_sales_y_test.shape

((800, 4), (800,), (200, 4), (200,))

In [59]:
# import RandomForestRegressor
from sklearn.ensemble import RandomForestRegressor
model = RandomForestRegressor()

In [61]:
# before processing using the model I need to transform the categories data into numeric
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer
# I need to transform Make, Colour, and Doors columns
categorical_features = ["Make", "Colour", "Doors"]
one_hot = OneHotEncoder()
transformer = ColumnTransformer([("hot_transform", one_hot, categorical_features)], remainder="passthrough")
# now we use the transformer to transform X data frame
trans_car_sales_X = transformer.fit_transform(car_sales_X)
trans_car_sales_X

array([[0.00000e+00, 1.00000e+00, 0.00000e+00, ..., 1.00000e+00,
        0.00000e+00, 3.54310e+04],
       [1.00000e+00, 0.00000e+00, 0.00000e+00, ..., 0.00000e+00,
        1.00000e+00, 1.92714e+05],
       [0.00000e+00, 1.00000e+00, 0.00000e+00, ..., 1.00000e+00,
        0.00000e+00, 8.47140e+04],
       ...,
       [0.00000e+00, 0.00000e+00, 1.00000e+00, ..., 1.00000e+00,
        0.00000e+00, 6.66040e+04],
       [0.00000e+00, 1.00000e+00, 0.00000e+00, ..., 1.00000e+00,
        0.00000e+00, 2.15883e+05],
       [0.00000e+00, 0.00000e+00, 0.00000e+00, ..., 1.00000e+00,
        0.00000e+00, 2.48360e+05]])

#### NOTE: THE TRANSFORMED RESULTS

This is a bit odd that the results are bunch of matrices of  numbers. Even the index 0 matrices are numbers. I can't give much sense of what it means.

In [62]:
trans_car_sales_X[0]

array([0.0000e+00, 1.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00,
       0.0000e+00, 0.0000e+00, 0.0000e+00, 1.0000e+00, 0.0000e+00,
       1.0000e+00, 0.0000e+00, 3.5431e+04])

In [67]:
car_sales_X.iloc[0]

Make             Honda
Colour           White
Odometer (KM)    35431
Doors                4
Name: 0, dtype: object

In [68]:
car_sales.head()

Make Colour  Odometer (KM) Doors  Price
0   Honda  White          35431     4  15323
1     BMW   Blue         192714     5  19943
2   Honda  White          84714     4  28343
3  Toyota  White         154365     4  13434
4  Nissan   Blue         181577     3  14043

#### NOTE: I NEED AN ALTERNATIVE

Maybe the best alternative will be using the pandas library called [pandas.get_dummies](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.get_dummies.html). 

One caveat though; all columns data frame input must be type object. Here is why I need to convert "doors" column to object

In [66]:
# convert the "Doors" data into type object
car_sales["Doors"] = car_sales["Doors"].astype(object)
# make the dummies
dummies = pd.get_dummies(car_sales[["Make", "Colour", "Doors"]])
dummies

Make_BMW  Make_Honda  Make_Nissan  Make_Toyota  Colour_Black  \
0           0           1            0            0             0   
1           1           0            0            0             0   
2           0           1            0            0             0   
3           0           0            0            1             0   
4           0           0            1            0             0   
..        ...         ...          ...          ...           ...   
995         0           0            0            1             1   
996         0           0            1            0             0   
997         0           0            1            0             0   
998         0           1            0            0             0   
999         0           0            0            1             0   

     Colour_Blue  Colour_Green  Colour_Red  Colour_White  Doors_3  Doors_4  \
0              0             0           0             1        0        1   
1              1             0           0             0        0        0   
2              0             0           0             1        0        1   
3              0             0           0             1        0        1   
4              1             0           0             0        1        0   
..           ...           ...         ...           ...      ...      ...   
995            0             0           0             0        0        1   
996            0             0           0             1        1        0   
997            1             0           0             0        0        1   
998            0             0           0             1        0        1   
999            1             0           0             0        0        1   

     Doors_5  
0          0  
1          1  
2          0  
3          0  
4          0  
..       ...  
995        0  
996        0  
997        0  
998        0  
999        0  

[1000 rows x 12 columns]

#### NOTE: This is the same order
The pandas.get_dummies data Frame arrangement is the same as the trans_car_sales_X arrangement.

Look at the first row of the dummies: the Make_Honda is the 2nd column and has a value of 1 the same as the trans_car_sales_X[0] 2nd value.

This is what transforming really does. It converts the data into dummies which have a boolean-like type. It only gives 1 to the suitable data (eg. Make_Honda, Color_White, Doors_4)

This is the best way to convert categorical data into numerics I guess. Although I still hope they will just give Make value between 0 to 3 to represent each car maker name.


In [72]:
# let's see the X again
car_sales_X["Make"].value_counts()

Toyota    398
Honda     304
Nissan    198
BMW       100
Name: Make, dtype: int64

In [74]:
# let's re-fitting the model 
# we need to re-split the model again though:
car_sales_X_train, car_sales_X_test, car_sales_y_train, car_sales_y_test = train_test_split(trans_car_sales_X, car_sales_y, test_size=0.2)

# remember we already set our RandomForestRegressor into model variable (object) thus I can just use model to fit
model.fit(car_sales_X_train, car_sales_y_train)

RandomForestRegressor()

In [75]:
model.score(car_sales_X_test, car_sales_y_test)

0.2023241292740252

TODO: PROCESSING MISSING VALUES!